---
title: "Assignment5_Code"
author: "Stephanie Rosen"
date: "12/12/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

# Description of Dataset

\#Functions

# Data Acquisition, Exploration, Filtering, and Quality Control

```{r Loading required packages, echo=FALSE}

library(tidyverse)
library(ggplot2)
library(Biostrings)
library(rentrez)
library(muscle)
library(DECIPHER)
```

We require two datasets for this project: (1) Community data for the wetlands and (2) Genetic data from NCBI

```{r community data acquistion, exploration, filtering and quality control}

#import dataset
df_Odonata <- PEI_NP_Wetlands_Odonata_2010_2018_data

names(df_Odonata)
dim(df_Odonata)

#total number of unique species?
#73
#but lots of errors with species names. We will remove NA values, any species names that do not follow proper scientific naming procedures, and also any where the species level is not indicated. 


#remove chunks of text in brackets but still have full species names
df_Odonata$Scientific.name <- str_remove(string = df_Odonata$Scientific.name, "\\(.+\\)" )

#There are some species that have a third unecessary word in their name. We will remove that too.
#df_Odonata[str_detect(string = df_Odonata$Scientific.name, "^[A-z]+ [A-z]+ [A-z]+$"),]

df_Odonata$Scientific.name <- str_remove(string = df_Odonata$Scientific.name, "Walker" )
df_Odonata$Scientific.name <- str_remove(string = df_Odonata$Scientific.name, "Say" )
df_Odonata$Scientific.name <- str_replace(string = df_Odonata$Scientific.name, "anax", "Anax")
df_Odonata$Scientific.name <- str_replace(string = df_Odonata$Scientific.name, "leucorrhinia", "Leucorrhinia")
df_Odonata$Scientific.name <- str_replace(string = df_Odonata$Scientific.name, "aeshna", "Aeshna")

#remove first row (French header), scientific names with NA values, irrelevant columns, and any names that do not follow. 
df_Odonata <- df_Odonata[-1,] %>%
  filter(!is.na(Scientific.name)) %>%
  select(!exuvia.observer.number & !Observer..order.number & !note) %>%
  filter(str_detect(Scientific.name, "^[A-z]+ [A-z]+$")) %>%
  filter(!str_detect(Scientific.name, "NOT ODONATA")) %>%
  filter(!str_detect(Scientific.name, "unknown Aniaoptera")) %>%
  filter(!str_detect(Scientific.name, "Enallagma sp"))

unique(df_Odonata$Scientific.name)

#Now that we have clean scientific names, let's look at the different ponds we have:
unique(df_Odonata$Survey.site)

#Ideas for looking into biodiversity
#pick one year and explore

df_sites <-
  df_Odonata %>%
  group_by(Year) %>%
  count(Survey.site)

df_counts_by_site <- 
  pivot_wider(df_sites, id_cols = Survey.site, names_from = Year, values_from = Year)

#from this table, we see that there were different sites measured every year. I chose to look at the sites that were measured continuously over the 10 year period. Let's look at the sites in common. That is the sites of Bell's pond, Bog Pond, and Bowley Pond

df_Odonata_common_sites <-
  df_Odonata %>%
  filter(Survey.site == "Bells Pond" | Survey.site == "Bog Pond" | Survey.site == "Bowley Pond")

#Now the df_Odonata_common sites data has all the data for the ponds measured every year. We are ready to explore the data.



```

Now we have clean scientific name data for our different communities. Let's explore the data

```{r data exploration}

#Exploring total number of counts at each pond
df_Odonata_counts_by_year_plot <- df_Odonata_common_sites %>%
  group_by(Survey.site) %>%
  count(Survey.site)

barplot(df_Odonata_counts_by_year_plot$n,
        names.arg = df_Odonata_counts_by_year_plot$Survey.site)

#Exploring total number of counts at each pond per year

df_Odonata_unique_by_year_pond <- df_Odonata_common_sites %>%
  group_by(Year, Survey.site) %>%
  summarise(n = n_distinct(Scientific.name))

ggplot(df_Odonata_unique_by_year_pond, 
       aes(fill=Survey.site, y=n, x=Year)) + 
  geom_bar(position="dodge", stat="identity") +
  labs(y = "Number of Unique Odonata Species", 
       title = "Number of Unique Odonata Species per Pond by Year") +
  scale_fill_manual(values = c("#D81B60","#1E88E5","#FFC107"), 
                    name = "Survey Site") +
  theme(plot.title = element_text(hjust = 0.5))



```

-we can see that there were a lot less species collected in 2014 -the biodiversity per year seems to vary quite a bit -because we are only interested in the biodiversity of the ponds, for this analysis we will ignore the species counts and leave them for a future analysis -we will investigate the unique species over the 10 year period, and then potentially for each year (maybe just 2008, 2012, and 2017?)


```{r putting data into community matrix format for future analysis}
df_Odonata_common_sites$count <- as.numeric(df_Odonata_common_sites$count)


test <- df_Odonata_common_sites %>%
  group_by(Survey.site,Scientific.name) %>%
  summarize()

#Using aggregate() function. What are we aggregating? latitude data (argument x). What are we grouping by? province_state (we need to make a list for the by argument). What function do we want to apply to the groups? mean, with na.rm set to TRUE. We are creating a new dataframe, dfAvgLat1.
dfAvgLat1 <- aggregate(x = dfBOLD$lat, by = list(ProvinceState = dfBOLD$province_state), FUN = function(x) mean(x, na.rm = TRUE))

tapply(X = df_Odonata_common_sites$count, 
       INDEX = list(df_Odonata_common_sites$Scientific.name, df_Odonata_common_sites$Survey.site), FUN = sum, simplify = T)

df_Odonata_spread <- pivot_wider(data = df_Odonata_common_sites,
                                 id_cols = Scientific.name,
                                 names_from = Survey.site,
                                 values_from = Survey.site)



```


Now that our data looks good, let's get those sequences.

```{r NCBI data}

#Let's get a list of unique species grouped by ponds

df_Odonata_unique_by_pond <- df_Odonata_common_sites %>% 
  group_by(Survey.site, Scientific.name) %>%
  summarise()
  

#now we can get NCBI data

COI_search_ids <- vector() #empty vector for all the search ids

unique_species <- unique(df_Odonata_common_sites$Scientific.name)
length(unique_species)

#loop through and get the first ID for each species COI sequence
for (species in unique_species) {
  COI_search <- entrez_search(db = "nuccore", term = paste(species, "[ORGN] AND COI[GENE] AND 650:665[SLEN]", sep = " "))
  COI_search_ids <- append(COI_search_ids, COI_search$ids[1])
  }

length(COI_search_ids) #perfect! We got an ID for each species
#but wait, some of the sequence id's are null
#we will have to remove these sequences from our analysis

COI_search_ids_filtered <- COI_search_ids[-c(15, 20, 26, 29)]

#Now let's download the actual sequences
COI_fetch <- entrez_fetch(db = "nuccore", id = COI_search_ids_filtered, rettype = "fasta")

head(COI_fetch)

#Writing  FASTA file to working directory
write(COI_fetch, "COI_fetch.fasta", sep = "\n") 

#Converting the FASTA file to a BioStrings DNAStringSet format
stringSet <- readDNAStringSet("COI_fetch.fasta")
class(stringSet)
head(names(stringSet))

#Put sequences into dataframe format
df_COI_Odonata <- data.frame(COI_Title = names(stringSet), COI_Sequence = paste(stringSet))

```

Now that we have obtained the sequences, we will filter them:

```{r filtering and quality control of sequences}

missing.data <- 0.01
length.var <- 50

df_COI_Odonata <- df_COI_Odonata %>%
  mutate(COI_Sequence2 = str_remove_all(COI_Sequence, "^N+|N+$|-")) %>%
  filter(str_count(COI_Sequence2, "N") <= (missing.data * str_count(COI_Sequence))) %>%
  filter(str_count(COI_Sequence2) >= median(str_count(COI_Sequence2)) - length.var & str_count(COI_Sequence2) <= median(str_count(COI_Sequence2)) + length.var)


#Cleaning up the species names to proper taxonomic names
df_COI_Odonata$Species_Name <- word(df_COI_Odonata$COI_Title, start = 2L, end = 3L)

remove(missing.data, length.var)
```





# Main Software Tools Description

# Main Analysis


```{r Sequence Alignment}

df_COI_Odonata$COI_Sequence2 <- DNAStringSet(df_COI_Odonata$COI_Sequence2)

#Naming the sequences before they are aligned
names(df_COI_Odonata$COI_Sequence2) <- df_COI_Odonata$Species_Name

#aligning sequences using default settings
dfCOI.alignment <- DNAStringSet(muscle::muscle(df_COI_Odonata$COI_Sequence2))

#writing sequence alignment to file in FASTA format
writeXStringSet(dfCOI.alignment, file = "Odonata.fas", format = "fasta")

#look at sequences in browser
BrowseSeqs(dfCOI.alignment)

#looks good!


```


```{r Phylogenetic Tree}



```


#Results and Discussion

#Acknowledgements

#References
